### PR TITLE
Soviet Soldier Volkov & Chitzkoi fully ported

### PR DESCRIPTION
**Pull request by AngryBirdz**
Hi there, I've just completed my porting of the Counterstrike expansion mission named "Soviet Soldier Volkov & Chitzkoi" (scu35ea). The map is fully operational but it might need some playtest by someone else than myself just in case there's still some bugs and / or balance issue that I did'nt noticed.

Have a nice day :)

### CODE REVIEW

**Comment by reaperrr** [comment]
It looks like our lua syntax checker can't handle the '&' in the folder name, maybe just write 'and' instead.

It would also be nice if you could list the lua scripts in https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.sln, some reviewers might use that for debugging.

**Comment by pchote** [comment]
The `&` character has special meaning (put job in background) on Linux and macOS, and so can't be used unescaped in file paths.  Yes, please replace it with `and` or something else.

**Comment by AngryBirdz** [comment]
Alright guys I did everything you asked and I think I did'nt forget anything.

UPDATE
-Added myself in the authors file (I forgot the first time)
-Removed the special character in the filename
-Listed my lua scripts in OpenRA.sln
-Removed two useless line in weapons.yaml (just a leftover of me fooling around with some stuff)

**Comment by AngryBirdz** [comment]
Can you explain to me why this has happened? I seriously don't know. I double checked my code and the comp "Goodguy" is listed both in the map.yaml and the lua script. What did I miss?

**Comment by pchote** [comment]
Player names are case sensitive.  See `GoodGuy` vs `Goodguy`.

File: AUTHORS [hunk 1]
```diff
@@ -104,6 +104,7 @@ Also thanks to:
     * Michael Rätzel
     * Michael Silber (frühstück)
     * Michael Sztolcman (s1w_)
+	* Mike Gagné (AngryBirdz)
```
Comment (pchote) [inline AUTHORS]: Please use 4 spaces here instead of a tab.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+2TNK:
+	Buildable:
+		Prerequisites: ~vehicles.allies, ~techlevel.medium
+        
+CA:
+	Buildable:
+		Prerequisites: ~syrd, ~techlevel.high
+        
+C2:
+	Inherits@AUTOTARGET: ^AutoTargetGroundAssaultMove
```
Comment (pchote) [inline mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: Have you tried defining `Inherits@2: ^ArmedCivilian` instead?  This would save a lot of duplication.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	Health:
+		HP: 250000
+	Armor:
+		Type: Heavy
+	-Crushable:
+	RevealsShroud:
+		Range: 10c0
+	Demolition:
+	Mobile:
+		-Voice: Move
```
Comment (pchote) [inline mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: Why do these voices need to be removed?

Comment (AngryBirdz) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: Because these voices can't be used with the GenericVoice voiceset. It make the game crash.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml (hunk 2 above)
Comment (AngryBirdz) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: Yeah I'll upload the new file soon

**Comment by Mailaender** [comment]
A very minor issue:

image

The mines display the multiplayer user name on hover.

**Comment by Mailaender** [comment]
Another polishing issue:

image

The lab (and hospital) still shows it's multiplayer function in the tooltip.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea-AI.lua [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	elseif (OreTruck01.IsDead and OreTruck02.IsDead) and greece.Resources <= 699 then
+		return
+	end
+
+	local delay = Utils.RandomInteger(DateTime.Seconds(7), DateTime.Seconds(10))
+	local toBuild = { Utils.Random(AlliedArmorTypes) }
+  local Rally = Utils.Random(AlliedWarFactRally)
+  Utils.Do(AlliedWarFact, function(fact)
+    fact.RallyPoint = Rally.Location
+  end)
```
Comment (Mailaender) [inline mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea-AI.lua]: Indention look wrong here.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea-AI.lua [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	end)
+end
+
+ProduceNavyGuard = function()
+	if NavalYard01.IsDead then
+		return
+	elseif (OreTruck01.IsDead and OreTruck02.IsDead) and greece.Resources <= 2399 then
+		return
+	end
+  NavalYard01.RallyPoint = waypoint26.Location
```
Comment (Mailaender) [inline mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea-AI.lua]: You seem to be using a mix of tabs and spaces for indention. Please just use tabs.

**Review by Mailaender — DISMISSED** [review]
Looking good otherwise. Please squash all fixup commits together when you are done.

**Comment by AngryBirdz** [comment]
I'm a noob at this. Can you explain to me what this "PR: Needs +2" label mean? Did I forgot to do something?

**Comment by Smittytron** [comment]
OpenRA requires two approval reviews to merge a PR. The tag just means there's one approval.

**Comment by AngryBirdz** [comment]
Ok thx for the info.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+			end
+			Trigger.RemoveFootprintTrigger(id)
+		end
+	end)
+
+	Trigger.OnKilled(RiflemanGuard01, function()
+		ProduceInfantry() --Greece will start infantry production once this unit is dead just like in the original mission
+	end)
+
+--Kill the ore trucks when the two allied refinery are destroyed
```
Comment (abcdefg30) [inline mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua]: Why is this necessary? (It looks a bit odd to me.)

**Review by abcdefg30 — COMMENTED** [review]
The tooltip of the alloy factory needs to be hidden/adjusted:
grafik

Otherwise, well done! The mission plays nicely ingame. Only issue I had: The allied forces often only damage the concrete walls of your base and you still get a "Our base is under attack" warning which I found to be  a bit annoying/distracting (for just concrete...). Do you think this should be changed?

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+				VolkovIsDead(a)
+			end
+			if a.Type == "zkoi" then
+				ChitzkoiIsDead(a)
+			end
+		end)
+		Media.PlaySpeechNotification(player, "ReinforcementsArrived")
+		transport.Paradrop(CPos.New(21, 82))
+	end)
+
```
Comment (abcdefg30) [inline mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua]: I forgot to mention: trailing white space. :)

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+			normal: Normal
+			hard: Hard
+		Default: easy
+
+Player:
+	PlayerResources:
+		DefaultCash: 0
+	LobbyPrerequisiteCheckbox@GLOBALBOUNTY:
+		Enabled: False
+		Locked: True
```
[the remark this answers is not in the document]
Comment (AngryBirdz) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: Are you sure that LobbyPrerequisiteCheckbox@GLOBALBOUNTY is pre-defined in campaign-rules.yaml? Because I don't see it.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua (hunk 6 above)
Comment (AngryBirdz) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua]: The vanilla mission work that way.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/map.yaml [hunk 9]
```diff
@@ -0,0 +1,1449 @@
...
+	PlayerReference@Neutral:
+		Name: Neutral
+		OwnsWorld: True
+		NonCombatant: True
+		Faction: england
+	PlayerReference@Creeps:
+		Name: Creeps
+		NonCombatant: True
+		Faction: england
+        Enemies: Greece, GoodGuy, Spain, France, USSR
```
[the remark this answers is not in the document]
Comment (AngryBirdz) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/map.yaml]: For the tooltip description i have two questions:
1) How do you make it show ingame? Can't find the setting.
2) Do you know wich trait point to it? Because I'm having a hard time finding it.
Thx!

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (pchote) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: @abcdefg30 Do we need it? Bounties default to off in the default game rules.

[the remark this answers is not in the document]
Comment (abcdefg30) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: No, we don't necessarily need this. However, we usually lock the options in the campaign missions.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/map.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (abcdefg30) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/map.yaml]: The tooltip should just appear when you hover with your mouse over the building. The trait is `TooltipDescription` iirc.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua (hunk 6 above)
Comment (abcdefg30) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua]: Can you please remove it (still)? I would consider that an improvement over the original mission. (It does not make sense to me why the ore trucks should blow up. As player, I would probably be rather confused by this.)

**Comment by AngryBirdz** [comment]
I tried to fix the Tooltip Description with the trait you told me to mod but the check failed somehow?

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+V01:
+	SpawnActorOnDeath:
+		Actor: healcrate
+        
+HOSP:
+	Tooltip:
+		GenericName: Hospital
+		GenericVisibility: Enemy, Ally, Neutral
+		GenericStancePrefix: false
+		TooltipDescription: Destroy this hospital to get healing equipment.
```
Comment (Mailaender) [inline mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: This is indented wrong. `TooltipDescription` is it's own trait:

```
	TooltipDescription:
		Description: Destroy this hospital to get healing equipment.
```

should work.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+	Buildable:
+		Prerequisites: ~disabled
+	Health:
+		HP: 400000
+	Tooltip:
+		Name: Alloy Facility
+		GenericName: Alloy Facility
+		GenericVisibility: Enemy, Ally, Neutral
+		GenericStancePrefix: false
+		TooltipDescription: This facility produce our stolen armor plating.
```
Comment (Mailaender) [inline mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: Same here.

Comment (abcdefg30) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: While you are at it, please change `produce` to `produces`.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml (hunk 10 above)
Comment (AngryBirdz) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: notworking
Not sure if by its own trait you mean outside of tooltip but if I'm doing it like that the game wont even load the map.

working
On the other hand if I indent it like you told me but inside tooltip the map load. Can you confirm this will work before I update the commit? Thanks you.

Comment (pchote) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: [TooltipDescription](https://github.com/OpenRA/OpenRA/blob/d33ac0c8384ae360b725b5886f4c28ade95af315/OpenRA.Mods.Common/Traits/TooltipDescription.cs) is a different trait to [Tooltip](https://github.com/OpenRA/OpenRA/blob/d33ac0c8384ae360b725b5886f4c28ade95af315/OpenRA.Mods.Common/Traits/Tooltip.cs).  When you indent the definition as in your second example the game tries to load a `TooltipDescription` *field* on the `Tooltip` trait, which does not exist and is therefore ignored.

The lint check (which you can view from the "Details" link in the "Some checks were not successful" panel below reports this more explicitly:
```
OpenRA.Utility(1,1): Error: rules.yaml:75 refers to a trait field `TooltipDescription` that does not exist on `Tooltip`.
OpenRA.Utility(1,1): Error: rules.yaml:101 refers to a trait field `TooltipDescription` that does not exist on `Tooltip`.
```

Comment (Mailaender) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: > Not sure if by its own trait you mean outside of tooltip but if I'm doing it like that the game wont even load the map.

How do you test this map? Your local OpenRA installation or this branch compiled from source?

Comment (AngryBirdz) [reply mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/rules.yaml]: My local OpenRA installation.

**Comment by AngryBirdz** [comment]
I've managed to fix it by using the playtest build. Just tell me if I forgot something.

File: mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+				if not unit.IsDead then
+					Trigger.OnIdle(unit, function(unit)
+						if unit.IsInWorld then
+							unit.Hunt()
+						end
+					end)
+				end
+			end)
+			if Map.LobbyOption("difficulty") == "hard" then
+				Trigger.ClearAll(RiflemanGuard01)
```
Comment (abcdefg30) [inline mods/ra/maps/soviet-soldier-volkov-n-chitzkoi/scu35ea.lua]: This crashes when `RiflemanGuard01` is already dead.

**Comment by AngryBirdz** [comment]
Ok I've fixed the crash related to RiflemanGuard01. Tell me if there's something else.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b8a3afc425a2588092f098e49b9e0f251723972e...0c6096c3c6d91c95ade6036751abe3481d52e9c5)
